patches/server/0887-Add-option-to-hide-entities-by-default.-This-option-.patch

```diff
@@ -0,0 +1,118 @@
+         this.origin = location.toVector();
+         this.originWorld = location.getWorld().getUID();
+@@ -1503,7 +1512,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
+     }
+ 
+     public boolean isSilent() {
+-        return (Boolean) this.entityData.get(Entity.DATA_SILENT);
++        return (Boolean) this.entityData.get(Entity.DATA_SILENT) || this.isHiddenEntityByDefault();  // Paper - don't play any sounds if this entity is hidden by default
+     }
+ 
+     public void setSilent(boolean silent) {
+@@ -3605,7 +3614,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
+     }
+ 
+     public boolean broadcastToPlayer(ServerPlayer spectator) {
+-        return true;
++        return !this.hideEntityByDefault; // Paper - only broadcast to a player if this entity is shown by default
+     }
+ 
+     @Override
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+index a92755211e3d42934b5efaa3f201c6c19ab7d2b4..2d7dcd08f0e93fa6bb40d2995a88b908e6c35c1b 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+@@ -1230,6 +1230,22 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+     // Spigot end
+ 
+     // Paper start
++
++    @Override
++    public boolean isHiddenByDefault() {
++        return getHandle().isHiddenEntityByDefault();
++    }
++
++    @Override
++    public void hideEntityByDefault() {
++        getHandle().setHideEntityByDefault(true);
++    }
++
++    @Override
++    public void showEntityByDefault() {
++        getHandle().setHideEntityByDefault(false);
++    }
++
+     @Override
+     public Location getOrigin() {
+         Vector originVector = this.getHandle().getOriginVector();
+@@ -1270,7 +1286,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
+     public boolean isInWaterOrBubbleColumn() {
+         return getHandle().isInWaterOrBubble();
+     }
+-    
++
+     public boolean isInWaterOrRainOrBubbleColumn() {
+         return getHandle().isInWaterRainOrBubble();
+     }
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index 6feeadad9aecb7d63e24d5daae115a93e39aeb3d..f861b695aac3c8633e3bb308cdda42fe5e7c5f06 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1638,7 +1638,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public boolean canSee(org.bukkit.entity.Entity entity) {
+-        return !this.hiddenEntities.containsKey(entity.getUniqueId());
++        return !this.hiddenEntities.containsKey(entity.getUniqueId()) || !entity.isHiddenByDefault();
```

Comment:
wouldn't you want an && here not || ?

Reply:
No, since an entity can be hidden overall or specifically hidden just for this player